### PR TITLE
Remove credentials and session from Onyx on sign out

### PR DESCRIPTION
**Pull request by luacmartins**
### Details
Removes session and credentials from local storage after sign out.

### Fixed Issues
$ https://github.com/Expensify/Expensify/issues/193459

### Tests
1. Login
2. Sign out
3. Refresh the page and verify that you still see the login page.

- [X] Verify that no errors appear in the JS console

### QA Steps
Steps above.

- [ ] Verify that no errors appear in the JS console

### Tested On

- [X] Web
- [ ] Mobile Web
- [X] Desktop
- [ ] iOS
- [ ] Android

#### Web

https://user-images.githubusercontent.com/22219519/151057676-81afb72e-a709-4fc8-be65-e97f9667efd1.mov

#### Desktop

https://user-images.githubusercontent.com/22219519/151057913-d23bdfeb-bd39-4f0a-a5bc-42f4e0c051b7.mov

### CODE REVIEW

**Review by danieldoglas — APPROVED** [review]
LGTM

File: src/libs/actions/Session/index.js [hunk 1]
```diff
@@ -83,6 +83,8 @@ function signOut() {
         })
             .catch(error => Onyx.merge(ONYXKEYS.SESSION, {error: error.message}));
     }
+    Onyx.set(ONYXKEYS.SESSION, null);
+    Onyx.set(ONYXKEYS.CREDENTIALS, null);
     Timing.clearData();
     redirectToSignIn();
```
Comment (kidroca) [inline src/libs/actions/Session/index.js:89]: Hey why is this necessary `redirectToSignIn` is clearing the whole storage anyway?

Comment (luacmartins) [reply src/libs/actions/Session/index.js:89]: For some reason `redirectToSignIn` was failing to clear the storage and discard the `authToken`. This PR solved an issue where after being redirected, the user could refresh the page and would be logged in again.

Comment (marcaaron) [reply src/libs/actions/Session/index.js:89]: >For some reason redirectToSignIn was failing to clear the storage and discard the authToken. This PR solved an issue where after being redirected, the user could refresh the page and would be logged in again.

What was the reason?

Comment (luacmartins) [reply src/libs/actions/Session/index.js:89]: Not sure yet. I'll investigate it as part of https://github.com/Expensify/Expensify/issues/194430

Comment (marcaaron) [reply src/libs/actions/Session/index.js:89]: Oh hmm maybe we can do something like "read from cache only while `Onyx.clear()` is working". I'm not really sure what the exact problem or solution would be here.

Comment (luacmartins) [reply src/libs/actions/Session/index.js:89]: >Maybe the problem is that it takes 2+ seconds to reach the line calling Storage.clear and not that clearing IDB storage is slow?

That makes sense. I just timed execution for `Onyx.clear` so it's possible that Storage.clear is fast but it takes a while to get there.

>Perhaps we can move the Storage.clear call to happen before notifying subscribers

I tried this but it resulted in the app automatically refreshing and the user "signing in" again. Why do we need to call `keyChanged` for every key to update components that will unmount?

Comment (kidroca) [reply src/libs/actions/Session/index.js:89]: > Why do we need to call `keyChanged` for every key to update components that will unmount?

I don't think that's the purpose we call it for 
We want to inform all active subscribers that the key value is now `null` 
Why do we **want** that? Maybe it matters for them, otherwise their props would remain unchanged - nothing would change `isAuthenticated` that's based on `props.session.authToken`

Most components will unmount because `isAuthenticated` would switch to `false` so they re-render in vain just before they unmount...

Comment (kidroca) [reply src/libs/actions/Session/index.js:89]: Here's another thought:

React is a tree - we have a root component and then branches, branches, branches
Our connections are likely following the tree hierarchy, connections at the start of the list are the ones of root components 

When a single key changes we iterate all subscribers (starting from root) and notify them of the change
But when there are multiple keys we iterate keys and for each iterate all subscribers

When we change many keys like in `.clear` it might be more performant to iterate subscribers and then keys for each subscriber 
This way root subscribers would unmount whole branches of components

Comment (kidroca) [reply src/libs/actions/Session/index.js:89]: Ideally we should be able to call a component `setState` with all the keys that changed in one go and then do the same for remaining child components

Comment (tgolen) [reply src/libs/actions/Session/index.js:89]: > When we change many keys like in .clear it might be more performant to iterate subscribers and then keys for each subscriber
This way root subscribers would unmount whole branches of components

This seems like a good concept, but I don't think I am fulling grasping what you mean (not sure that I need to). Does it mean that if I have component A that is subscribed to three keys, then when `clear()` is called, component A re-renders three times?

Another possible alternative is something like adding a `clearWithoutTriggeringSubscribers()` type method. I think ideally, I'd rather make it more performance like @kidroca is exploring.

Comment (kidroca) [reply src/libs/actions/Session/index.js:89]: > Does it mean that if I have component A that is subscribed to three keys, then when `clear()` is called, component A re-renders three times?

yes and no, depending on what you consider a render 
yes - it would trigger `render` 3 times and any also trigger any children `render`s.
no - the screen might not re-render at all, or only do it once  

setState is called per key and each time triggers a re-render 
```js
subscriber.withOnyxInstance.setState({
                    [subscriber.statePropertyName]: data,
                });
```

> Another possible alternative is something like adding a `clearWithoutTriggeringSubscribers()` type method.

This would not make any subscribers aware of the change and `isAuthenticated` would not transition to `false`

Comment (tgolen) [reply src/libs/actions/Session/index.js:89]: > This would not make any subscribers aware of the change and isAuthenticated would not transition to false

Yep, you're right.

OK, so another thought. We know that when we are signing out, we want everything to get unmounted. What if we unmount everything before calling `clear()`? Practically, that could be something like:

1. Click sign out
2. Go to /signout which is just a special stand-alone page (everything else in the app would be unmounted at this point)
3. Clear out Onyx
4. Redirect to sign in

Maybe that's still too much of a hack. Again, I'd rather fix the performance in the first place, but I did want to see what other alternatives we could think of.

Comment (kidroca) [reply src/libs/actions/Session/index.js:89]: The best alternative (besides making clear or Onyx faster) is already applied

```js
    Onyx.set(ONYXKEYS.SESSION, null);
    Onyx.set(ONYXKEYS.CREDENTIALS, null);
```

Removes `authToken` unmounting 95% of App and takes us to the SignIn page
Then Onyx.clear clears the rest of the stored values 

@luacmartins Onyx.clear happens faster with these changes and doesn't take 2.3s, right?

Comment (kidroca) [reply src/libs/actions/Session/index.js:89]: It seems JS blocking time is 100ms
Maybe in the past it was 1000 for some reason, or it was never  the cause for the delay...

It might depend on the amount of data stored in index DB
Or if reads / writes were happening while clearing for some reason
